compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangParserListener.java

```diff
@@ -866,9 +866,30 @@ public void exitInclusiveRecordTypeDescriptor(BallerinaParser.InclusiveRecordTyp
 
         boolean isAnonymous = !(ctx.parent.parent instanceof BallerinaParser.FiniteTypeUnitContext);
 
+        // Validate type AB "A"| record { string f; };
+        if (!isAnonymous) {
+            isAnonymous = checkIfAnonymousInTypeDef(ctx);
+        }
+
         this.pkgBuilder.addRecordType(getCurrentPos(ctx), getWS(ctx), isAnonymous, false, false);
     }
 
+    /**
+     * Validate if this is a anonymous record type describe in union with a type descriptor.
+     *
+     * @param ctx the current context to be validated. can be InclusiveRecordTypeDescriptor or ExclusiveRecordTypeDes.
+     * @return true if is part of a union type descriptor with on the fly definition
+     */
+    private boolean checkIfAnonymousInTypeDef(ParserRuleContext ctx) {
+        if (!(ctx instanceof BallerinaParser.InclusiveRecordTypeDescriptorContext ||
```

Comment:
Do we need to check the ctx type?
 I feel like we are pretty sure about this from the places we call this logic.